sklearn/semi_supervised/_self_training.py

```diff
@@ -0,0 +1,338 @@
+        self.threshold = threshold
+        self.selection_criterion = selection_criterion
+        self.n_best = n_best
+        self.max_iter = max_iter
+        self.verbose = verbose
+
+    def fit(self, X, y):
+        """
+        Fits this ``SelfTrainingClassifier`` to a dataset.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+            array representing the data
+
+        y : array-like, shape = (n_samples,)
+            array representing the labels. Unlabeled samples should have the
+            label -1.
+
+        Returns
+        -------
+        self : object
+            returns an instance of self.
+        """
+        X, y = check_X_y(X, y)
+
+        if self.base_classifier is None:
+            raise ValueError("base_classifier cannot be None!")
+
+        self.base_classifier_ = clone(self.base_classifier)
+
+        if self.max_iter is not None and self.max_iter < 0:
+            msg = "max_iter must be >= 0 or None, got {}".format(self.max_iter)
+            raise ValueError(msg)
+
+        if not 0 <= self.threshold < 1:
+            msg = "threshold must be in [0,1), got {}".format(self.threshold)
+            raise ValueError(msg)
+
+        if self.selection_criterion not in ['threshold', 'n_best']:
+            raise ValueError("selection_criterion must be either 'threshold' "
+                             "or 'n_best', got "
+                             "{}".format(self.selection_criterion))
+
+        if y.dtype.kind in ['U', 'S']:
+            raise ValueError("y has dtype string. If you wish to predict on "
+                             "string targets, use dtype object, and use -1"
+                             " as the label for unlabeled samples.")
+
+        has_label = y != -1
+
+        if np.all(has_label):
+            warnings.warn("y contains no unlabeled samples", UserWarning)
+
+        if self.selection_criterion == 'n_best' and (self.n_best > X.shape[0] -
+                                                     np.sum(has_label)):
+            warnings.warn("n_best is larger than the amount of unlabeled "
+                          "samples. All unlabeled samples will be labeled in "
+                          "the first iteration", UserWarning)
+
+        self.transduction_ = np.copy(y)
+        self.labeled_iter_ = np.full_like(y, -1)
+        self.labeled_iter_[has_label] = 0
+
+        self.n_iter_ = 0
+
+        while not np.all(has_label) and (self.max_iter is None or
+                                         self.n_iter_ < self.max_iter):
+            self.n_iter_ += 1
+            self.base_classifier_.fit(
+                X[safe_mask(X, has_label)],
+                self.transduction_[safe_mask(self.transduction_, has_label)])
+
+            if self.n_iter_ == 1:
+                _validate_estimator(self.base_classifier)
+
+            # Predict on the unlabeled samples
+            prob = self.base_classifier_.predict_proba(
```

Comment:
With n_best should we also support decision_function here (at least in the binary case)? Maybe later.